src/simulators/statevector/qubitvector.hpp

```diff
@@ -2080,6 +2114,53 @@ double QubitVector<data_t, Derived>::expval_pauli(const reg_t &qubits,
   return std::real(apply_reduction_lambda(std::move(lambda), (size_t) 0, (data_size_ >> 1)));
 }
 
+/*******************************************************************************
+ *
+ * MULTI-PAULI
+ *
+ ******************************************************************************/
+template <typename data_t, typename Derived>
+void QubitVector<data_t, Derived>::apply_multipauli(const reg_t &qubits, const std::string &pauli){
+  uint_t x_mask, z_mask, num_y, x_max;
+  std::tie(x_mask, z_mask, num_y, x_max) = pauli_masks_and_phase(qubits, pauli);
+
+  // Special case for only I Paulis
+  if (x_mask + z_mask == 0) {
+    return;
+  }
+  std::complex<data_t> phase;
+  compute_phase(num_y, phase);
+  const uint_t DIM = 1ULL << qubits.size();
+
+  // specialize x_max == 0
+  if (!x_mask) {
+    auto lambda = [&](const int_t i)->void {
+        if (z_mask && (AER::Utils::popcount(i & z_mask) & 1)) {
+             data_[i] *= -1;
+        }
+        data_[i] *= phase;
+    };
+    apply_lambda(lambda);
+    return;
+  }
+
+  const uint_t mask_u = ~MASKS[x_max + 1];
+  const uint_t mask_l = MASKS[x_max];
+  auto lambda = [&](const int_t i)->void {
+    int_t idxs[2];
+    idxs[0] = ((i << 1) & mask_u) | (i & mask_l);
+    idxs[1] = idxs[0] ^ x_mask;
+    for (int_t j = 0; j < 2; ++j) {
+      if (z_mask && (AER::Utils::popcount(idxs[j] & z_mask) & 1)) {
+        data_[idxs[j]] *= -1;
+      }
+      data_[idxs[j]] *= phase;
+    }
+    std::swap(data_[idxs[0]], data_[idxs[1]]);
+  };
+  apply_lambda(lambda, (size_t) 0, (data_size_ >> 1));
+}
+
 //------------------------------------------------------------------------------
```

Comment:
There is still quite a lot of duplication between `pauli_expval` and `apply_multipauli`, any way to address it?

Reply:
Good question. Currently the duplicated code is mostly bit tricks which do not encapsulate nicely; let's see what happens when we add Pauli measurements.

Reply:
Pauli measurements are more similar to expectation value than gates. I'm probably going to rewrite `expval_pauli` as `pauli_expval_and_measure`, and it will be essentially the same function as now, with small changes.

Reply:
I main difference between Pauli expval and Pauli measure is expval is deterministic and can be computed as a single reduction without storing an intermediate probability vector.

With measure you need to sample an outcome from the Pauli measurement distribution (which could also be computed as a reduction).